### ISSUE TITLE
TypeError: name must be a str, not a MagicMock

### ISSUE THREAD

**Issue by buxx** [pnuckowski/aioresponses#117]
Hello,

I have a Strange issue when use aioresponses mocking. I can't reproduce with example script and i can't show the concerned code. So i just hope error can suggest you something.

``` python
>       with aioresponses() as mock:

tests/xxxxxxxxxxxxx.py:570: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv3.7/lib/python3.7/site-packages/aioresponses/core.py:203: in __enter__
    self.start()
venv3.7/lib/python3.7/site-packages/aioresponses/core.py:238: in start
    self.patcher.start()
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:1386: in start
    result = self.__enter__()
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:1336: in __enter__
    _name=self.attribute, **kwargs)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:2238: in create_autospec
    **kwargs)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:1842: in __init__
    _safe_super(MagicMixin, self).__init__(*args, **kw)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:936: in __init__
    _spec_state, _new_name, _new_parent, **kwargs
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:394: in __init__
    self._mock_add_spec(spec, spec_set, _spec_as_instance, _eat_self)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:451: in _mock_add_spec
    _spec_as_instance, _eat_self)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/unittest/mock.py:90: in _get_signature_object
    return func, inspect.signature(sig_func)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/inspect.py:3075: in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/inspect.py:2825: in from_callable
    follow_wrapper_chains=follow_wrapped)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/inspect.py:2282: in _signature_from_callable
    return _signature_from_function(sigcls, obj)
../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/inspect.py:2166: in _signature_from_function
    kind=_VAR_POSITIONAL))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("_name") raised in repr()] Parameter object at 0x7f81827e0120>
name = <MagicMock name='_request.__code__.co_varnames.__getitem__()' id='140194217530704'>
kind = <_ParameterKind.VAR_POSITIONAL: 2>

    def __init__(self, name, kind, *, default=_empty, annotation=_empty):
        try:
            self._kind = _ParameterKind(kind)
        except ValueError:
            raise ValueError(f'value {kind!r} is not a valid Parameter.kind')
        if default is not _empty:
            if self._kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
                msg = '{} parameters cannot have default values'
                msg = msg.format(_get_paramkind_descr(self._kind))
                raise ValueError(msg)
        self._default = default
        self._annotation = annotation
    
        if name is _empty:
            raise ValueError('name is a required attribute for Parameter')
    
        if not isinstance(name, str):
            msg = 'name must be a str, not a {}'.format(type(name).__name__)
>           raise TypeError(msg)
E           TypeError: name must be a str, not a MagicMock

../../.pythonz/pythons/CPython-3.7.1/lib/python3.7/inspect.py:2483: TypeError
```

You can close issue if error say Nothing to you. I go back here if solution found ...

**Comment by buxx** [comment]
Solution found, hope it can help:

`with aioresponses() as mock:` was done after an other `with aioresponses() as mock:`. So aioresponses try to mock `'aiohttp.client.ClientSession._request'` second time and fail like indicated bellow.

**Comment by lermana** [comment]
I got a similar error when called a `@patch`'d function from another `@patch`'d function ... particularly, the same method was `@patch`'d for both of these, and I think (maybe due to generator internals?) that didn't seem to play. I refactored to ensure one testing function wasn't calling another with the same methods `@patch`'d, and the problem was fixed.

I think [this](https://github.com/python/cpython/blob/3.9/Lib/unittest/mock.py#L1334) is the code that's causing the issue (the generator context manager):

```
with self.decoration_helper(
                              patched,
                              args,
                              keywargs) as (newargs, newkeywargs):
      return func(*newargs, **newkeywargs)
````

Thanks for your comment, @buxx - it helped me figure out how to get around this. I guess this is the intended behavior, because you want `patch` to traverse all the way down. That said, maybe this error could be messaged better? I.e. instead of:
> TypeError: name must be a str, not a MagicMock

it could prove helpful to wrap that context in a `try / except` and throw a message more like (if the above is encountered):
> Cannot patch the same object twice within the scope of the same function.
